### ISSUE TITLE
Inconsistent Slicing of ArrayData for StructArray

### ISSUE THREAD

**Issue by tustvold** [apache/arrow-rs#1750]
**Problem**

ArrayData::Slice contains a special case for StructArray where it recurses the offset into its children. However, it preserves the offset on the parent ArrayData, in order for the validity buffer to work correctly.

The result is the offset has been pushed down to the child arrays, but is still present on the parent ArrayData. This makes it unclear if an offset should or should not be applied to the child_array.

**Proposal**

There are longer term suggestions around handling offsets in ArrayData differently, but until then I would like to propose:

* Remove the ArrayData::slice special-case
* Slice child data within `StructArray` when constructing `boxed_fields`

This is consistent with how we handle offsets elsewhere, with ArrayData a dumb container, and the `Array` providing the logic to interpret the offset correctly.

Thoughts @nevi-me

**Comment by HaoYang670** [comment]
cc @nevi-me @alamb @viirya

**Comment by tustvold** [comment]
> And it is not clear whether we have pushed the new offset to children

The suggestion is never to push offsets down within ArrayData, and so you should never have this ambiguity? 

When constructing an instance of StructArray you "hydrate" the offsets to boxed fields to avoid having to recompute slices on access, but this does not modify the underlying ArrayData?

**Comment by HaoYang670** [comment]
>The suggestion is never to push offsets down within ArrayData

Although we don't push down the offset within `ArrayData::slice` (Based on your 1st suggestion `Remove the ArrayData::slice special-case`), we do push it down within `StructArray::from` (Based on your 2nd suggestion `Slice child data within StructArray when constructing boxed_fields`) because we must slice on each `child_data` to construct the correct child array.

As a result, I guess your suggestions (slicing child arrays within `StructArray::from()`) and the current implementation (slicing the child data within `ArrayData::slice()`) are equivalent. I am not 100% sure.

**Comment by tustvold** [comment]
Aah, I see the confusion. I'm suggesting storing slices of the child data as boxed_fields on StructArray. So in StructArray::from it would inspect the ArrayData, compute slices, and store those as members. I am not suggesting StructArray::from make any modifications to ArrayData.

**Comment by HaoYang670** [comment]
> So in StructArray::from it would inspect the ArrayData, compute slices, and store those as members.
Implement as this:
```rust
    fn from(data: ArrayData) -> Self {
        let boxed_fields = data
            .child_data()
            .iter()
            .map(|cd| make_array(cd.slice(data.offset(), data.len())))
            .collect();

        Self { data, boxed_fields }
    }
```
?

**Comment by tustvold** [comment]
I _think_ that should work, but there could be some arcane edge case I've forgotten.

**Comment by HaoYang670** [comment]
This will not modify the parent ArrayData, which is great. 
But It will cause the inconsistency between `bixed_fields` and `self.data.child_data`, which means `StructArray::column(self, 0) !=  make_array(self.data.child_data()[0])`?

**Comment by HaoYang670** [comment]
I don't find `StructArray::slice` in the c++ implementation (I just find the ArrayData::slice https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/data.cc#L94-L110)
And in Arrow2, everything is pushed down to Buffer and Bitmap

**Comment by tustvold** [comment]
>  But It will cause the inconsistency between bixed_fields and self.data.child_data, which means StructArray::column(self, 0) != make_array(self.data.child_data()[0])?

We have the same "inconsistency" on other array types, e.g. `GenericStringArray::value_offsets` or `PrimitiveArray::values`. I personally don't think this is a problem.

> I don't find StructArray::slice in the c++ implementation

Yeah this is just a usability thing, the Rust implementation is just calling through to `make_array(self.data_ref().slice(offset, length))`.

> everything is pushed down to Buffer and Bitmap

I would definitely support this change, and it would be the better long-term solution. It has been discussed at length, all that really remains is someone to volunteer to do it :sweat_smile:. The only outstanding sticking point has been handling offsets for BooleanArrays, but I don't imagine there would be resistance to special casing these as a Bitmap in ArrayData...

**Comment by alamb** [comment]
cc @jhorstmann and @bjchambers who I think have interest in this area

**Comment by jhorstmann** [comment]
Sorry for being late to this discussion, the proposed change makes sense to me. If I understand it correctly, this would be a silent behavior change of the `column`/`columns`/`columns_ref` methods on `StructArray`, which previously did not take the struct offset into account. We might want to deprecate those and instead add new methods like `pub fn column_slice(&self, pos: usize) -> &ArrayRef` (exact name up for discussion).

**Comment by tustvold** [comment]
> which previously did not take the struct offset into account.

Is this not a bug?

**Comment by tustvold** [comment]
> The trouble was that if we don't pass down the offset and length to the ArrayData of a's children, we'd be bound to always knowing a's offset, which forces us to compute it each time we access a or any of its children.

I think caching the sliced array on `StructArray::boxed_fields` would avoid this, and is consistent with what we do for PrimtiveArray / StringArray, etc... ?

> but yea perhaps first-prize would be propagating offsets and value lengths to a redesigned Buffer and Bitmap

Agreed, and I think the first step of this is removing the remaining places that "leak" offsets into the Array-level interfaces :smile:

**Comment by tustvold** [comment]
Closing in favor of removing the offsets entirely (#1799)

**Comment by tustvold** [comment]
Closed by #4061
